### DISCUSSION TITLE
CoreWCF.ConfigurationManager too dependency-heavy?

### DISCUSSION THREAD

**Discussion by fbenoit-coexya** [CoreWCF/CoreWCF#1380] [Ideas]
Hi,
Before .NET Core threw a stone in it, I had managed to use ConfigurationChannelFactory and its server equivalent to put all our WCF configuration in the same file, no longer scattered between our multiple servers and clients (which made it much easier to resolve a conflict).

Unfortunately, having a shared configuration file has become much harder:
- If I want my configuration file to contain a `<system.serviceModel>` section that CoreWCF can read, it means every project that wants to read from this file needs to know what a `CoreWCF.Configuration.ServiceModelSectionGroup` is, as well as all sections in it.
- This means every project that wants to read the file needs to reference CoreWCF.ConfigurationManager
- And, at least in the .NET Standard 2.0 version, referencing CoreWCF.ConfigurationManager references the whole Microsoft.AspNetCore infrastructure! Adding the package added 70+ other packages to the project.

Having a smaller package with just what is needed to parse the configuration file would help a lot.

**Comment by mconnew** [comment]
With the current design that's not possible as CoreWCF.ConfigurationManager has code to instantiate each of the supported bindings. At some future time I plan to get service config working using IConfiguration, at which point CoreWCF.ConfigurationManager would simply provide an IConfigurationProvider (or some similar interface in that space) to convert the app.config entries into IConfiguration items so wouldn't need to directly instantiate any class. Until that happens, CoreWCF.ConfigurationManager basically has to depend on the world.  

Something that might help you, CoreWCF.ConfigurationManager is able to read a separate config file which isn't the main web.config/app.config file. You can break out the \<system.serviceModel\> config into a separate file of its own.

**Reply by fbenoit-coexya** [reply]
No way to isolate the `ConfigSection`s and enums into their own, separate package?
Looks like I have not choice then, and will indeed have to give CoreWCF its own, dedicated config file. But this also means separating the client config from the server config once again, which once again opens the window for incompatible binding configurations. :-(

**Reply by mconnew** [reply]
Could you do something at runtime or compile time to strip it out? .NET doesn't auto load the app.config file at process launch like it used to. It's more like a third party configuration library where it only loads the file when you use ConfigurationManager to instantiate a Configuration instance. At runtime you could load the config file into something like XDocument, strip out the system.serviceModel node, then write it back to disk with a different name. Then point ConfigurationManager to the new file.  

At build time, you could use [Microsoft.VisualStudio.SlowCheetah](https://github.com/Microsoft/slow-cheetah) to use an xslt transform at build time to strip out the `<system.serviceModel>` nodes from the config file. This way you can have a common config for multiple projects, and some projects strip out the CoreWCF pieces, and others leave them in.

**Reply by fbenoit-coexya** [reply]
That sounds doable. It will require saving the stripped configuration file to some writable location (or doing so compile-time) but it beats putting them in separate files to start with.
Thanks!

**Comment by fbenoit-coexya** [comment]
In the end, stripping the configuration file with SlowCheetah caused issues with some projects that have both client and server parts, so instead I:

- Duplicated most of the XxxxSection/XxxxElement code from CoreWCF to my own project (so a unique code independent of the CoreWCF / ASP.NET Core boat anchor could read the file)
- On the "server" side only, I load with the lightweight code then I convert it back to a  `CoreWCF.Configuration.ServicesSection` and `CoreWCF.Configuration.BindingsSection` by round-tripping it though `ConfigurationSection.SerializeSection()` and `ConfigurationSection.DeserializeSection()`. 
  - I could do this easily for `ServicesSection` by extending the class, 
  - but `BindingsSection` being sealed required me to use Reflection to break encapsulation :-( It's possible I missed something (like, there could be public-but-low-level-enough code in `System.Configuration` that calls DeserializeSection()) but so far it's all I have.
